contao/classes/BackendUser.php

```diff
@@ -223,22 +226,32 @@ public function authenticate()
 			return true;
 		}
 
-		list($path) = explode('?', \Environment::get('request'), 2);
+		/** @var KernelInterface $kernel */
+		global $kernel;
+
+		$container = $kernel->getContainer();
+		$route     = $container->get('request_stack')->getCurrentRequest()->attributes->get('_route');
 
-		if (substr($path, -13) == '/contao/login')
+		if ($route == 'contao_backend_login')
 		{
 			return false;
 		}
 
-		$strRedirect = 'contao/';
+		$parameters = [];
 
 		// Redirect to the last page visited upon login
-		if (substr($path, -7) == '/contao' || substr($path, -15) == '/contao/preview')
+		if ($route == 'contao_backend' || $route == 'contao_backend_preview')
 		{
-			$strRedirect .= '?referer=' . base64_encode(\Environment::get('request'));
+			$parameters['referer'] = base64_encode(\Environment::get('request'));
 		}
 
-		\Controller::redirect($strRedirect);
```

Comment:
That's not yet available in this branch, the exception handling is a follow-up to this.